### PR TITLE
solve on non-conforming meshes & error estimators

### PR DESCRIPTION
**Pull request by Huangzizhou**

### CODE REVIEW

**Comment by Huangzizhou** [comment]
Compilation error on Mac, still finding the reason

File: src/mesh2D/NCMesh2D.hpp [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+{
+
+struct ArrayHasher2D 
+{
+    long operator()(const Eigen::Vector2i& a) const {
+        return (long)((long)984120265 * a[0] + (long)125965121 * a[1]);
+    }
+};
+
+struct ArrayHasher3D
```
Comment (teseoch) [inline src/mesh2D/NCMesh2D.hpp:15]: remove it

File: src/mesh2D/NCMesh2D.hpp [hunk 2]
```diff
@@ -0,0 +1,349 @@
+#pragma once
+
+#include <polyfem/Mesh2D.hpp>
+
+namespace polyfem
+{
+
+struct ArrayHasher2D
```
Comment (teseoch) [inline src/mesh2D/NCMesh2D.hpp:8]: move inside the class as private stuff

File: src/assembler/ElementAssemblyValues.cpp [hunk 3]
```diff
@@ -199,6 +199,12 @@ namespace polyfem
 		compute(el_index, is_volume, quadrature.points, basis, gbasis);
 	}
 
+	void ElementAssemblyValues::compute(const bool is_volume, const ElementBases &basis, const ElementBases &gbasis)
```
Comment (teseoch) [inline src/assembler/ElementAssemblyValues.cpp:202]: what is this?

File: src/assembler/ElementAssemblyValues.cpp [hunk 4]
```diff
@@ -349,6 +355,156 @@ namespace polyfem
 
 	}
 
+	void ElementAssemblyValues::compute(const bool is_volume, const Eigen::MatrixXd &pts, const ElementBases &basis, const ElementBases &gbasis)
```
Comment (teseoch) [inline src/assembler/ElementAssemblyValues.cpp:358]: and this?

File: src/basis/FEBasis2d.cpp [hunk 5]
```diff
@@ -762,33 +885,53 @@ int polyfem::FEBasis2d::build_bases(
 			b.bases[j].eval_basis(uvdx, vdx);
 			b.bases[j].eval_basis(uvdy, vdy);
 
-			assert((grad.col(0) - (vdx - val) / 1e-6).norm() < 1e-4);
-			assert((grad.col(1) - (vdy - val) / 1e-6).norm() < 1e-4);
+			// assert((grad.col(0) - (vdx - val) / 1e-6).norm() < 1e-4);
```
Comment (teseoch) [inline src/basis/FEBasis2d.cpp:888]: why remove assertions?

File: torus.msh [hunk 6]
```diff
@@ -0,0 +1,700 @@
+$MeshFormat
```
Comment (teseoch) [inline torus.msh:1]: don't add meshes to the repo. use the data repo

**Review by teseoch — CHANGES_REQUESTED** [review]

**Comment by teseoch** [comment]
can you also apply the formatting to all files?

**Comment by Huangzizhou** [comment]
> can you also apply the formatting to all files?

I don't understand this one

File: src/assembler/ElementAssemblyValues.cpp (hunk 3 above)
Comment (Huangzizhou) [reply src/assembler/ElementAssemblyValues.cpp]: removed

File: src/assembler/ElementAssemblyValues.cpp (hunk 4 above)
Comment (Huangzizhou) [reply src/assembler/ElementAssemblyValues.cpp]: removed

File: src/basis/FEBasis2d.cpp (hunk 5 above)
Comment (Huangzizhou) [reply src/basis/FEBasis2d.cpp]: reverted

File: torus.msh (hunk 6 above)
Comment (Huangzizhou) [reply torus.msh]: removed

**Review by teseoch — APPROVED** [review]

**Pull request merged**